### PR TITLE
Remove Tag creation for ports and add description for it

### LINKED ISSUE #543: disableServerTags: true is not working
/kind bug

**What steps did you take and what happened:**
I'm trying to provision Kubernetes Cluster on Top of our OpenStack Provider, which it doesn't support port tags.
I've disabled the server tags `disableServerTags: true` in the `OpenStackCluster` config and I have no tags in the `OpenStackMachineTemplate`

**What did you expect to happen:**
Create OpenStack cluster without tags  

**Anything else you would like to add:**
```
$ kubectl describe openstackmachine devops-control-plane-trf55
Name:         devops-control-plane-trf55
Namespace:    default
Labels:       cluster.x-k8s.io/cluster-name=devops
              cluster.x-k8s.io/control-plane=
              kubeadm.controlplane.cluster.x-k8s.io/hash=724735395
Annotations:  <none>
API Version:  infrastructure.cluster.x-k8s.io/v1alpha3
Kind:         OpenStackMachine
Metadata:
  Creation Timestamp:  2020-04-06T14:29:27Z
  Finalizers:
    openstackmachine.infrastructure.cluster.x-k8s.io
  Generation:  1
  Owner References:
    API Version:           controlplane.cluster.x-k8s.io/v1alpha3
    Kind:                  KubeadmControlPlane
    Name:                  devops-control-plane
    UID:                   d22ddafb-17f8-4c79-ad6d-5b285a48a335
    API Version:           cluster.x-k8s.io/v1alpha3
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Machine
    Name:                  devops-control-plane-n88ww
    UID:                   db838279-5626-4e23-8126-7c33b6640e02
  Resource Version:        2889
  Self Link:               /apis/infrastructure.cluster.x-k8s.io/v1alpha3/namespaces/default/openstackmachines/devops-control-plane-trf55
  UID:                     d62c98e5-c9ff-40f0-a91e-5052f3299a8d
Spec:
  Cloud Name:  otc_doa
  Clouds Secret:
    Name:       devops-cloud-config
    Namespace:  default
  Flavor:       s2.large.1
  Image:        Standard_CentOS_7_prev_packer_test
  Security Groups:
    Name:        devops-k8s-master
    Name:        devops-k8s-master
    Name:        devops-local-ssh
    Name:        devops-k8s-ssh
    Name:        devops-k8s
  Ssh Key Name:  kubernetes-devops
Status:
  Error Message:  OpenStack instance cannot be created: error creating Openstack instance: tagging port for server err: Resource not found
  Error Reason:   UpdateError
Events:           <none>
```

**Environment:**

- Cluster-api version: v0.3.3 
- Minikube/KIND version: v1.17.0
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #543

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jichenjc** [comment]
after PR, the port looks like:

```
             |
| binding:vnic_type     | normal                                                                                                                    |
| created_at            | 2020-07-24T14:41:36Z                                                                                                      |
| description           | Created by cluster-api-provider-openstack cluster capi-openstack-3                                                        |
.....
| status                | ACTIVE                                                                                                                    |
| tags                  |                                                                                                                           |
```

**Comment by hidekazuna** [comment]
If we really remove server tag at all, we have to update templates and api/v1alpha3/openstackmachine_types.go.

**Comment by jichenjc** [comment]
> machineTags

I only want to remove port tags ... for machines/servers, I want to keep them unless we can reach agreement ..

**Comment by hidekazuna** [comment]
> > > > machineTags variable is create in Remove L81 -L94, and used also in L198 - L202, and L206 - L215.
> > > > Remove all lines, please.
> 
> I only want to remove port tags ... for machines/servers, I want to keep them unless we can reach agreement .. so for api definition, same as well, unless we are sure we want to remove **all** tags
> we can modify there, for now, the PR is to handle #543 only , @hidekazuna please let me know your thoughts

Now I caught up the discussion in the issue and thought [@sbueringer 's comment](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/543#issuecomment-610745978) is the agreement.

**Comment by jichenjc** [comment]
ok, let me check that :)

**Comment by jichenjc** [comment]
@hidekazuna I updated the PR and the tags are removed now

**Comment by hidekazuna** [comment]
> @hidekazuna I updated the PR and the tags are removed now

Updated PR still has `disableServerTags`. The agreement is that we remove `disableServerTags` , and only add tags when they are explicitly defined in the cluster or machine resource.

**Comment by jichenjc** [comment]
@hidekazuna please help to check again, thanks

**Comment by jichenjc** [comment]
L190: Can trunk attribute be replaced with empty machineTags?  which L 190? I don't know the trunk is related to tags..
L226: Can server with empty machineTags create successfully? ===> yes, empty tag is allowed, my test showed I can create a cluster (the machines) without any tag

**Comment by hidekazuna** [comment]
> L190: Can trunk attribute be replaced with empty machineTags? which L 190? I don't know the trunk is related to tags..

Leave trunk port as is. 

> L226: Can server with empty machineTags create successfully? ===> yes, empty tag is allowed, my test showed I can create a cluster (the machines) without any tag

Thanks, understand.

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by hidekazuna** [comment]
@jichenjc Why do you recheck repeatedly? 
Please address my comment https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/596#pullrequestreview-460415786 :  This PR removes tags of network and router, but it is better similar to the instance; if Tags of OpenStackClusterSpec are added, add them to the network and router.

**Comment by jichenjc** [comment]
ok， need go back double check with those .. need focus 
anyway it's Friday :)

**Comment by jichenjc** [comment]
> OpenStackClusterSpec

@hidekazuna please check the logic 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/services/networking/network.go#L92

it's dynamic created network, we don't want someone to use our capo to create independent network , correct ? we only want to create capo own and known networks while them are automatically created and user is unaware of the network ... so how can they tag it ? 
basically, I think for networks capo automatically created, we previously tag it and we can add description to it later on ,for capo not related network, we take it as input (for example, the external network for floating IP) but we don't want to modify it 

so I think seems no need to make your proposed changes

**Comment by jichenjc** [comment]
keep recheck is because 

```
2020-08-07 13:28:32.407718 | ubuntu-bionic-large | [13:28:32] Total machines : 2 / Running : 1 .. waiting for 10 seconds
2020-08-07 13:28:42.966606 | ubuntu-bionic-large | NAME                                    PROVIDERID                                         PHASE
2020-08-07 13:28:42.967857 | ubuntu-bionic-large | capi-quickstart-control-plane-n7jbq     openstack://22f27e60-13a7-47d4-986b-592be31a0156   Running
2020-08-07 13:28:42.968026 | ubuntu-bionic-large | capi-quickstart-md-0-67b79f9848-5v6tq   openstack://eb48b147-9c49-46f0-8e7f-6e0a23f71a68   Provisioning
2020-08-07 13:28:43.234245 | ubuntu-bionic-large | cluster did not start in 30 mins ... bailing out!
```

seem this is not related to the PR

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
@hidekazuna please check my comments above and let me know your opinion, thanks

**Comment by hidekazuna** [comment]
> 
> 
> @jichenjc Why do you recheck repeatedly?
> Please address my comment #596 (review) : This PR removes tags of network and router, but it is better similar to the instance; if Tags of OpenStackClusterSpec are added, add them to the network and router.
> 
> @hidekazuna please check the logic
> https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/cloud/services/networking/network.go#L92
> 
> it's dynamic created network, we don't want someone to use our capo to create independent network , correct ? we only want to create capo own and known networks while them are automatically created and user is unaware of the network ... so how can they tag it ?
> basically, I think for networks capo automatically created, we previously tag it and we can add description to it later on ,for capo not related network, we take it as input (for example, the external network for floating IP) but we don't want to modify it
> 
> so I think seems no need to make your proposed changes

I understand you propose to delete tags for network, subnet, and router which is created by CAPO.
I proposed to add tags specified by the manifest in openStackCluster for network, subnet, and router which is created by CAPO.
Even if no network, subnet and router in the project, CAPO user can create multiple kubernetes cluster in the project and need to differentiate them.

**Comment by jichenjc** [comment]
ok, so you are suggesting to add tags for all the resource we created? 
e.g create a tag in the openstack cluster, if it's not specified, then do nothing, if it's given, then use it to tag all resources such as network/subnet ?

**Comment by hidekazuna** [comment]
> ok, so you are suggesting to add tags for all the resource we created?
> e.g create a tag in the openstack cluster, if it's not specified, then do nothing, if it's given, then use it to tag all resources such as network/subnet ?

Yes, at least current tagged resources. I respect current implementation

**Comment by jichenjc** [comment]
so 

```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
kind: OpenStackCluster
metadata:
  name: ${CLUSTER_NAME}
spec:
  cloudName: ${OPENSTACK_CLOUD}
  cloudsSecret:
    name: ${CLUSTER_NAME}-cloud-config
    namespace: ${NAMESPACE}
  managedAPIServerLoadBalancer: true
  managedSecurityGroups: true
  nodeCidr: 10.6.0.0/24
  dnsNameservers:
  - ${OPENSTACK_DNS_NAMESERVERS}
  disablePortSecurity: false
  tags: 
  -  xxxx
  -   yyy
```

all resource created by CPO will be taged by xxxx and yyyy, correct?

**Comment by hidekazuna** [comment]
@jichenjc Yes, I agree.

**Comment by jichenjc** [comment]
recheck

File: pkg/cloud/services/networking/network.go [hunk 1]
```diff
@@ -128,13 +128,12 @@ func (s *Service) ReconcileNetwork(clusterName string, openStackCluster *infrav1
...
-			clusterName,
-		}}).Extract()
-	if err != nil {
-		return err
+	if len(openStackCluster.Spec.Tags) > 0 {
+		_, err = attributestags.ReplaceAll(s.client, "networks", network.ID, attributestags.ReplaceAllOpts{
+			Tags: openStackCluster.Spec.Tags}).Extract()
+		if err != nil {
+			return err
+		}
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/network.go:136]: Since Tags is added to type.go, we need to reflect it to openStackCluster.Status.Network, openStackCluster.Status.Network.Subnet respectively.

**Review by hidekazuna — CHANGES_REQUESTED** [review]

File: pkg/cloud/services/networking/network.go (hunk 1 above)
Comment (jichenjc) [reply pkg/cloud/services/networking/network.go:136]: no, I need remove that part, because tag is given by `Spec`, no need to keep in `Status`..

**Comment by wangxiyuan** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

File: pkg/cloud/services/networking/network.go (hunk 1 above)
Comment (hidekazuna) [reply pkg/cloud/services/networking/network.go:136]: Since `Status` is the current status of the resources, it not odd that `openStackCluster.Status.Network` and `openStackCluster.Status.Network.Subnet` have tags.

Comment (jichenjc) [reply pkg/cloud/services/networking/network.go:136]: don't fully understand the comment, can you help to elaborate? thanks

Comment (hidekazuna) [reply pkg/cloud/services/networking/network.go:136]: I understand I can not add tags to Status as I expected. The followings are I tried.
At first I tried adding Tags with omitempty in types.go, but never tags are added to Status. Maybe the reason is  https://github.com/kubernetes-sigs/kubebuilder/issues/521.
Second, If added Tags without omitempty in types.go, tags with empty string to Status is not permitted and ended with error.
